### ISSUE TITLE
[typescript] strokeColor type only accepts Color, but code also accepts a string

### ISSUE THREAD

**Issue by dobesv** [paperjs/paper.js#1665]
# Description/Steps to reproduce

Try to assign `strokeColor` on an item to a `string` value.

# Link to reproduction test-case

https://github.com/dobesv/paperjs-ts-repro/blob/master/strokecolor-string.ts

# Expected result

Assigning a `string` value to `strokeColor` in typescript code does not result in a type error.

# Additional information

Because paper accepts a string value for `Color`, the typescript definitions should reflect that capability, and code that assigns a string value to `strokeColor` or other properties that do this conversion should type check without errors.

**Comment by sasensi** [comment]
The difficulty here, and it applies to many other cases, is that TypeScript does not offer a way to differentiate getter and setter types and the library allows setting value from different types whereas getting it will always return the same type (in this case a `Color`).
So I don't see any perfect solution.
- If we let as it is currently: `fillColor: Color`, we have got proper type assumption when we do for example `item.fillColor.red` but we have got a problem when doing `item.fillColor = 'red'`.
- If we change for `fillColor: Color | string`, `item.fillColor = 'red'` works but `item.fillColor.red` doesn't because the compiler thinks that `item.fillColor` could also be a `string`...

We are not the only one to encounter this problem as this thread shows https://github.com/Microsoft/TypeScript/issues/2521 and it doesn't seem that TypeScript has planned a solution for it.

@lehni, even if you are not using TypeScript, do you have an opinion on this ?

**Comment by dobesv** [comment]
`fillColor` I think can actually be `null` if the shape isn't filled, can it not?  Or does assigning `null` get converted to a transparent `Color` ?  If the latter, I guess I can just assign the transparent color myself.

**Comment by dobesv** [comment]
Yeah, it's an issue.  I guess I was used to the approach used in `@types/paper` which just falsely claims that you can get `null | string` back from this field, even though I guess in reality the field is always converted to a `Color`.

**Comment by anion155** [comment]
I think better approach would be to define getter and setters separately like I made in @types/paper for some props:
```typescript
strokeColor: Color | null;
getStrokeColor(): Color | null;
setStrokeColor(value: Color | string | null): void;
```

**Comment by sasensi** [comment]
@anion155, yes, this would be a good workaround in `TypeScript` context but I don't know if `Paper.js` users are supposed to use getter and setter methods directly as they are not documented.
Is it something that we can generalize to all gettable and settable properties or does this need to be documented differently ?
@lehni, what do you think ?

**Comment by lehni** [comment]
The getters/setters aren’t part of the public API and may go away in the future

**Comment by sasensi** [comment]
Ok, so I would say that this workaround is not suitable as it might mislead users.

**Comment by Domvel** [comment]
Just a few more infos about this issue.
`'get' and 'set' accessor must have the same type.`

[See the issue on Stackblitz](https://stackblitz.com/edit/typescript-get-set-type-issue?embed=1&file=index.ts)

[TypeScript Github Issue](https://github.com/Microsoft/TypeScript/issues/2521)

**Comment by MuTsunTsai** [comment]
I just migrated from `@types/paper` to the official d.ts lately and immediately having the same issue. In my opinion, TypeScript probably never will (and in fact shouldn't) allow different-typed getters and setters, as this obviously violates the core concept of what getters and setters are -- together they are supposed to create a illusion of a property, and since properties clearly can only have one type, so should getters and setters.
Therefore the proper workaround for this issue, in the true spirit of TypeScript, should be replacing all usage of `item.fillColor = 'red'` by `item.fillColor = new paper.Color('red')`.
